### PR TITLE
fix(payments): INT-3140 Enable checkout button with googlepay when use store credit - StripeV3

### PR DESCRIPTION
**Pull request by BC-AOrtiz**
[INT-3140](https://jira.bigcommerce.com/browse/INT-3140)

## What?
- Validate if shopper use store credit to enable the checkout button whit wallet

## Why?
- Shopper can place a order when use store credit and googlepay is enabled.

## Testing / Proof
Unit test

@bigcommerce/checkout

### CODE REVIEW

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx [hunk 1]
```diff
@@ -204,13 +207,23 @@ class WalletButtonPaymentMethod extends Component<
             method,
         } = this.props;
 
-        if (normalizeWalletPaymentData(method.initializationData)) {
+        if (normalizeWalletPaymentData(method.initializationData) || this.isStoreCreditUsed()) {
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx:210]: I think this calc is already in the state `checkoutState.data.isPaymentDataRequired`.
```suggestion
        if (normalizeWalletPaymentData(method.initializationData) || !isPaymentDataRequired) {
```

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx [hunk 2]
```diff
@@ -281,22 +294,28 @@ function mapFromCheckoutProps(
     { checkoutService, checkoutState }: CheckoutContextProps,
     { method }: WalletButtonPaymentMethodProps
 ): WithCheckoutWalletButtonPaymentMethodProps | null {
-    const { data: { getBillingAddress, getCheckout } } = checkoutState;
+    const { data: { getBillingAddress, getCheckout, getCustomer } } = checkoutState;
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx:297]: I think this is all you need:
```suggestion
    const { data: { getBillingAddress, getCheckout, isPaymentDataRequired } } = checkoutState;
```

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx [hunk 3]
```diff
@@ -281,22 +294,28 @@ function mapFromCheckoutProps(
...
+
     return {
         ...walletPaymentData,
         // FIXME: I'm not sure how this would work for non-English names.
         cardName: walletPaymentData && [billingAddress.firstName, billingAddress.lastName].join(' '),
         isPaymentSelected: some(checkout.payments, { providerId: method.id }),
         signOut: checkoutService.signOutCustomer,
+        isStoreCreditApplied: checkout.isStoreCreditApplied,
+        grandTotal: checkout.grandTotal,
+        storeCreditAmount,
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx:318]:
```suggestion
        isPaymentDataRequired
```

**Review by mauricio-sg — CHANGES_REQUESTED** [review]
I have a doubt and a couple of suggestions you could try. Let me know if you have any questions.

File: src/app/payment/Payment.tsx [hunk 4]
```diff
@@ -320,6 +320,7 @@ class Payment extends Component<PaymentProps & WithCheckoutPaymentProps & WithLa
 
         try {
             await applyStoreCredit(useStoreCredit);
+            await this.componentDidMount();
```
[the remark this answers is not in the document]
Comment (BC-AOrtiz) [reply src/app/payment/Payment.tsx:323]: I need to reload the payments to run again the validation to enable or disable the place order button when you click on use store credit checkbox

**Comment by BC-AOrtiz** [comment]
♻️ @mauricio-sg changes have been addressed, plase take another look

**Comment by ValeriaRivas** [comment]
💚  Tested

File: src/app/payment/Payment.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply src/app/payment/Payment.tsx:323]: So `componentDidUpdate` is the right place to check that, it also calls `toggleSubmit` method like `componentDidMount` does but this time validates if something changed, in this case you could check if `isPaymentDataRequired !== prevIsPaymentDataRequired` to determine if `toggleSubmit` should be called.

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx [hunk 5]
```diff
@@ -36,6 +36,7 @@ interface WithCheckoutWalletButtonPaymentMethodProps {
     expiryMonth?: string;
     expiryYear?: string;
     isPaymentSelected: boolean;
+    isPaymentDataRequired: boolean;
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx:39]: We try to keep props sorted, `isPaymentDataRequired` goes between `expiryYear` and `isPaymentSelected` sorry, my mistake, I didn't mention it before. 🙏

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx [hunk 6]
```diff
@@ -297,6 +299,7 @@ function mapFromCheckoutProps(
         cardName: walletPaymentData && [billingAddress.firstName, billingAddress.lastName].join(' '),
         isPaymentSelected: some(checkout.payments, { providerId: method.id }),
         signOut: checkoutService.signOutCustomer,
+        isPaymentDataRequired: isPaymentDataRequired(),
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.tsx:302]: Same here, `isPaymentDataRequired` goes after `cardName` 🙈

**Review by mauricio-sg — CHANGES_REQUESTED** [review]
Just a few more adjustments.

**Comment by BC-AOrtiz** [comment]
♻️ @mauricio-sg changes have been addressed, plase take another look

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx [hunk 7]
```diff
@@ -145,12 +145,28 @@ describe('WalletButtonPaymentMethod', () => {
     });
 
     it('disables submit button', () => {
-        mount(<WalletButtonPaymentMethodTest { ...defaultProps } />);
+        const props = merge( {}, defaultProps, {
+            isPaymentDataRequired: true,
+        });
+        mount(<WalletButtonPaymentMethodTest { ...props } />);
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx:151]: Just add the following code on line 50 and these changes won't be necessary.
```
        jest.spyOn(checkoutState.data, 'isPaymentDataRequired')
            .mockReturnValue(true);
```

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx [hunk 8]
```diff
@@ -145,12 +145,28 @@ describe('WalletButtonPaymentMethod', () => {
...
 
         expect(paymentContext.disableSubmit)
             .toHaveBeenCalledWith(defaultProps.method, true);
     });
 
+    it('enable submit button with store credit', () => {
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx:157]: 🍹
```suggestion
    it('enables submit button if payment data is not required', () => {
```

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx [hunk 9]
```diff
@@ -145,12 +145,28 @@ describe('WalletButtonPaymentMethod', () => {
...
+        const props = merge( {}, defaultProps, {
+            isPaymentDataRequired: false,
+        });
+
+        mount(<WalletButtonPaymentMethodTest
+            { ...props }
+        />);
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx:164]:
```suggestion
        jest.spyOn(checkoutState.data, 'isPaymentDataRequired')
            .mockReturnValue(false);

        mount(<WalletButtonPaymentMethodTest { ...defaultProps } />);
```

**Review by mauricio-sg — CHANGES_REQUESTED** [review]
These are the last changes, I promise 😬

**Comment by BC-AOrtiz** [comment]
♻️ @mauricio-sg changes have been addressed, plase take another look

File: src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx [hunk 10]
```diff
@@ -151,6 +154,18 @@ describe('WalletButtonPaymentMethod', () => {
             .toHaveBeenCalledWith(defaultProps.method, true);
     });
 
+    it('enables submit button if payment data is not required', () => {
+        jest.spyOn(checkoutState.data, 'isPaymentDataRequired')
+            .mockReturnValue(false);
+
+        mount(<WalletButtonPaymentMethodTest
+            { ...defaultProps }
+        />);
```
Comment (mauricio-sg) [inline src/app/payment/paymentMethod/WalletButtonPaymentMethod.spec.tsx:163]: 🍹 
```suggestion
        mount(<WalletButtonPaymentMethodTest { ...defaultProps } />);
```

**Review by mauricio-sg — APPROVED** [review]
Thank you @BC-AOrtiz, LGTM 👍

**Review by capsula4 — APPROVED** [review]

**Pull request merged**